pkg/credentials/provider_test.go

```diff
@@ -51,6 +52,79 @@ func TestNewProviderAssignsSessionDuration(t *testing.T) {
 	require.Equal(t, nonZeroSessionDuration, provider.SessionDuration)
 }
 
+func TestRetrieveSetsExpirationOnCredentials(t *testing.T) {
+	var audience, roleArn string
+	var sessionDuration = time.Second
+
+	jwtSource := mocks.JWTSource{}
+	defer jwtSource.AssertExpectations(t)
+	jwtSource.
+		On("FetchToken", mock.Anything).
+		Return("token", nil)
+
+	expiration := time.Now()
+	copyOfExpiration := expiration
+
+	credentials := sts.Credentials{
+		AccessKeyId:     aws.String(""),
+		Expiration:      &copyOfExpiration,
+		SecretAccessKey: aws.String(""),
+		SessionToken:    aws.String(""),
+	}
+
+	assumeRoleResponse := sts.AssumeRoleWithWebIdentityOutput{
+		Credentials: &credentials,
+	}
+
+	stsClient := mocks.STSAPI{}
+	defer stsClient.AssertExpectations(t)
+	stsClient.
+		On("AssumeRoleWithWebIdentityWithContext", mock.Anything, mock.Anything).
+		Return(&assumeRoleResponse, nil)
+
+	stsProvider := StaticSTSProvider(&stsClient)
+
+	provider, err := NewProvider(audience, roleArn, &jwtSource, sessionDuration, stsProvider)
+	require.NoError(t, err)
+
+	_, err = provider.Retrieve()
+	require.NoError(t, err)
+
+	require.EqualValues(t,
+		expiration.Round(0).Add(-time.Minute),
+		provider.Expiry.ExpiresAt())
+}
```

Comment:
I got a little confused by what was going on here until I read through the AWS source code. The expiration "window" that we supply isn't the raw window value, it's a delta that's used to shorten (not extend) the actual expiration window. So, if we supply a 1-minute expiration and the actual expiration time is 30 minutes, the net expiration time is now 29 minutes.

Reply:
Oh weird! So we need a negative window value to expand the actual window.